common/service/apml/apml.c

```diff
@@ -0,0 +1,388 @@
+apml_buffer apml_resp_buffer[APML_RESP_BUFFER_SIZE];
+
+uint8_t apml_read_byte(uint8_t bus, uint8_t addr, uint8_t offset, uint8_t *read_data)
+{
+	uint8_t retry = 5;
+	I2C_MSG msg;
+	msg.bus = bus;
+	msg.target_addr = addr;
+	msg.tx_len = 1;
+	msg.rx_len = 1;
+	msg.data[0] = offset;
+	int ret = i2c_master_read(&msg, retry);
+
+	if (ret) {
+		return APML_ERROR;
+	}
+	*read_data = msg.data[0];
+	return APML_SUCCESS;
+}
+
+uint8_t apml_write_byte(uint8_t bus, uint8_t addr, uint8_t offset, uint8_t write_data)
+{
+	uint8_t retry = 5;
+	I2C_MSG msg;
+	msg.bus = bus;
+	msg.target_addr = addr;
+	msg.tx_len = 2;
+	msg.rx_len = 0;
+	msg.data[0] = offset;
+	msg.data[1] = write_data;
+
+	if (i2c_master_write(&msg, retry)) {
+		return APML_ERROR;
+	}
+	return APML_SUCCESS;
+}
+
+static bool wait_HwAlert_set(apml_msg *msg, uint8_t retry)
+{
+	uint8_t read_data = 0;
+	for (uint8_t i = 0; i < retry; i++) {
+		if (!apml_read_byte(msg->bus, msg->target_addr, SBRMI_STATUS, &read_data)) {
+			if (read_data & 0x80) {
+				return true;
+			}
+		}
+		k_msleep(WAIT_TIME_MS);
+	}
+	return false;
+}
+/****************** MCA *********************/
+
+static uint8_t write_MCA_request(apml_msg *msg)
+{
+	uint8_t retry = 5;
+	I2C_MSG i2c_msg;
+	i2c_msg.bus = msg->bus;
+	i2c_msg.target_addr = msg->target_addr;
+	i2c_msg.tx_len = 9;
+	i2c_msg.data[0] = 0x73;
+	i2c_msg.data[1] = 0x07;
+	i2c_msg.data[2] = 0x08;
+	i2c_msg.data[3] = 0x86;
+	memcpy(&i2c_msg.data[4], msg->WrData, sizeof(mca_WrData));
+
+	if (i2c_master_write(&i2c_msg, retry)) {
+		return APML_ERROR;
+	}
+	return APML_SUCCESS;
+}
+
+static uint8_t read_MCA_response(apml_msg *msg)
+{
+	uint8_t retry = 5;
+	I2C_MSG i2c_msg;
+	i2c_msg.bus = msg->bus;
+	i2c_msg.target_addr = msg->target_addr;
+	i2c_msg.tx_len = 1;
+	i2c_msg.rx_len = 10;
+	i2c_msg.data[0] = 0x73;
+	if (i2c_master_read(&i2c_msg, retry)) {
+		return APML_ERROR;
+	}
+
+	memcpy(msg->RdData, &i2c_msg.data[1], sizeof(mca_RdData));
+	return APML_SUCCESS;
+}
+
+static uint8_t access_MCA(apml_msg *msg)
+{
+	if (write_MCA_request(msg)) {
+		printf("[%s] write MCA request failed.\n", __func__);
+		return APML_ERROR;
+	}
+
+	if (!wait_HwAlert_set(msg, RETRY_MAX)) {
+		printf("[%s] HwAlert not be set, retry %d times.\n", __func__, RETRY_MAX);
+	}
+
+	if (read_MCA_response(msg)) {
+		printf("[%s] read response failed.\n", __func__);
+		return APML_ERROR;
+	}
+
+	if (apml_write_byte(msg->bus, msg->target_addr, SBRMI_STATUS, 0x80)) {
+		printf("[%s] clear HwAlert failed.\n", __func__);
+		return APML_ERROR;
+	}
+	return APML_SUCCESS;
+}
+
+/****************** CPUID *******************/
+
+static uint8_t write_CPUID_request(apml_msg *msg)
+{
+	uint8_t retry = 5;
+	I2C_MSG i2c_msg;
+	i2c_msg.bus = msg->bus;
+	i2c_msg.target_addr = msg->target_addr;
+	i2c_msg.tx_len = 10;
+	i2c_msg.data[0] = 0x73;
+	i2c_msg.data[1] = 0x08;
+	i2c_msg.data[2] = 0x08;
+	i2c_msg.data[3] = 0x91;
+	memcpy(&i2c_msg.data[4], msg->WrData, sizeof(cpuid_WrData));
```

Comment:
verify msg is not NULL when passed into function

Reply:
Will check if msg is NULL.